awscli/customizations/cloudformation/parse_sub.py

```diff
@@ -0,0 +1,125 @@
+"""
+This module parses CloudFormation Sub strings.
+
+For example:
+
+    !Sub abc-${AWS::Region}-def-${Foo}
+
+The string is broken down into "words" that are one of four types:
+
+    String: A literal string component
+    Ref: A reference to another resource or paramter like ${Foo}
+    AWS: An AWS pseudo-parameter like ${AWS::Region}
+    GetAtt: A reference to an attribute like ${Foo.Bar}
+"""
+#pylint: disable=too-few-public-methods
+
+DATA = ' '  # Any other character
+DOLLAR = '$'
+OPEN = '{'
+CLOSE = '}'
+BANG = '!'
+
+class WordType:
```

Comment:
Nit picky here but could be `class WordType(Enum)`.  https://docs.python.org/3/library/enum.html.

or to freeze it so the values can't be changed.
```python
@dataclass(frozen=True)
class WordType
```